### ISSUE TITLE
miner gives zero weights

### ISSUE THREAD

**Issue by z1w** [KevinMusgrave/pytorch-metric-learning#276]
https://github.com/KevinMusgrave/pytorch-metric-learning/blob/19133d928cdc0aa4d1b49ddf30d5f9c81198649b/src/pytorch_metric_learning/miners/distance_weighted_miner.py#L44

I was using this distance-weighted miner, sometimes it fails because the returned weights are all zeros. Printing out all the tensors inside, I found that the log_weights, after exponentiation are almost all zeros except for the diagonal one, whose distance were cutoff to 0.5, the distances of other entries in the distance matrix after clamp are 0.7-0.9. 

After `weights = torch.exp(log_weights - torch.max(log_weights[~inf_or_nan]))`, the weights matrix becomes almost all zeros except for the diagonal but the diagonal are the same input points so they will be masked out. ` weights * mask` will lead to a matrix with all 0s. This happens under several different scenarios.

**Comment by KevinMusgrave** [comment]
It's possible for all the weights to be zero if all the negative distances are greater than ```nonzero_loss_cutoff```. Another possible scenario is if there is only 1 class in the batch.

It's also possible for the miner to return 0 triplets if there are no positive pairs in the batch.

**Comment by z1w** [comment]
I think it's none of them because the mask matrix is neither all zeros nor all ones and `mat < self.nonzero_loss_cutoff` returns an all true matrix.

Here is the link to the four tensors as inputs to the mine function: https://drive.google.com/drive/folders/1ttYiEYNIPTXgCE9Y6C608sYMB2GvnSPX?usp=sharing

The naming is consistent with the naming of the mine function inputs.

**Comment by KevinMusgrave** [comment]
Thanks for the samples. Are you getting 0 triplets with that input? I'm getting 150:

![image](https://user-images.githubusercontent.com/2314891/107066452-fe648080-67ab-11eb-9f24-42c71dca40ee.png)

**Comment by z1w** [comment]
Mine returns a different result. The miner function will trigger assert error caused by the all zero weight matrix, which will be used sampleMultinomial function. Let me update the package to see if the issue can be resolved. Thank you for checking this out.

**Comment by z1w** [comment]
Hi, I have updated the package and the weights are nonzero for the embedding tensor and labels tensor I referenced earlier today. However, as I run the training, the similar situation happened. The weight tensor becomes zero and after normalization the tensor becomes all nan.

The example tensors are still in the https://drive.google.com/drive/folders/1ttYiEYNIPTXgCE9Y6C608sYMB2GvnSPX?usp=sharing directory. If I ran the script above, I will have torch.Size([0]) as the output. As far as I can tell, none of the situations mentioned above applies here.

**Comment by KevinMusgrave** [comment]
I see the problem. As the embedding dimensionality grows, ```log_weights``` grows as well. I'm not sure if this computational problem was addressed in the paper, and my implementation is actually based off of theirs. They used 128-dim embeddings, and you're using 1024-dim embeddings.

I'm not sure how to fix this at the moment, but here's a workaround for now:

Cast the embeddings to ```torch.float64```:
![image](https://user-images.githubusercontent.com/2314891/107107804-78732480-6801-11eb-8902-39f810961e28.png)

**Comment by KevinMusgrave** [comment]
One possibility:

I could add an ```__init__``` option to calculate weights based on a normal distribution
![image](https://user-images.githubusercontent.com/2314891/107107913-31d1fa00-6802-11eb-8916-c8ec3eb2f5cb.png)

**Comment by z1w** [comment]
Thanks, that's exactly what I am trying to do, to see how embedding size can influence the performance. It seems that even if casting to float64, for higher dimension, those small numbers are going to 0 at some point. The normal distribution based sampling can be viewed as uniformly sample the negative examples on the sphere. I don't really know its effect on training. I can experiment with it to collect some empirical evidence.

**Comment by z1w** [comment]
Or add an alternative routine checking if the weight tensors are nan's, then use normal distribution based sampling. This might be better.

**Comment by KevinMusgrave** [comment]
I might not get around to fixing this until Wednesday, so if you figure something out, feel free to make a pull request to merge the changes.

**Comment by KevinMusgrave** [comment]
@z1w I couldn't figure out a way to fix the high dimensionality problem, so I made ```UniformHistogramMiner```, which you can find in this [pull request](https://github.com/KevinMusgrave/pytorch-metric-learning/pull/280).

**Comment by KevinMusgrave** [comment]
I think ```UniformHistogramMiner``` is a good alternative to ```DistanceWeightedMiner``` so I'm closing this issue.
